src/poly/evaluator.rs

```diff
@@ -0,0 +1,542 @@
+        let index = self.polys.len();
+        self.polys.push(poly);
+
+        AstLeaf {
+            index,
+            rotation: Rotation::cur(),
+            _evaluator: PhantomData::default(),
+        }
+    }
+
+    /// Evaluates the given polynomial operation against this context.
+    pub(crate) fn evaluate(
+        &self,
+        ast: &Ast<E, F, B>,
+        domain: &EvaluationDomain<F>,
+    ) -> Polynomial<F, B>
+    where
+        E: Copy + Send + Sync,
+        F: FieldExt,
+        B: BasisOps,
+    {
+        // Traverse `ast` to collect the used leaves.
+        fn collect_rotations<E: Copy, F: Field, B: Basis>(
+            ast: &Ast<E, F, B>,
+        ) -> HashSet<AstLeaf<E, B>> {
+            match ast {
+                Ast::Poly(leaf) => vec![*leaf].into_iter().collect(),
+                Ast::Add(a, b) | Ast::Mul(AstMul(a, b)) => {
+                    let lhs = collect_rotations(a);
+                    let rhs = collect_rotations(b);
+                    lhs.union(&rhs).cloned().collect()
+                }
+                Ast::Scale(a, _) => collect_rotations(a),
+                Ast::LinearTerm(_) | Ast::ConstantTerm(_) => HashSet::default(),
+            }
+        }
+        let leaves = collect_rotations(ast);
+
+        // Produce the rotated polynomials.
+        let rotated: HashMap<_, _> = leaves
+            .iter()
+            .cloned()
+            .map(|leaf| {
+                (
+                    leaf,
+                    B::rotate(domain, &self.polys[leaf.index], leaf.rotation),
+                )
+            })
+            .collect();
+
+        // We're working in a single basis, so all polynomials are the same length.
+        let poly_len = self.polys.first().unwrap().len();
+        // Check the level of parallelization we have available.
+        let num_threads = multicore::current_num_threads();
+        // We scale the number of chunks by a constant factor, to ensure that if not all
+        // threads are available, we can achieve more uniform throughput and don't end up
+        // waiting on a couple of threads to process the last chunks.
+        let num_chunks = num_threads * 4;
+        // Calculate the ideal chunk size for the desired throughput. We use ceiling
+        // division to ensure the minimum chunk size is 1.
+        //     chunk_size = ceil(poly_len / num_chunks)
+        let chunk_size = (poly_len + num_chunks - 1) / num_chunks;
+        // Now re-calculate num_chunks from the actual chunk size.
+        //     num_chunks = ceil(poly_len / chunk_size)
+        let num_chunks = (poly_len + chunk_size - 1) / chunk_size;
+
+        // Split each rotated polynomial into chunks.
+        let chunks: Vec<HashMap<_, _>> = (0..num_chunks)
+            .map(|i| {
+                rotated
+                    .iter()
+                    .map(|(leaf, poly)| {
+                        (
+                            *leaf,
+                            poly.chunks(chunk_size)
```

Comment:
It's inefficient to waste a thread on a small final chunk. We should instead split the poly more evenly, i.e. with some chunks having ceil(poly_len / num_chunks) elements and some having floor(poly_len / num_chunks).